File: backend/backend/queries/timetable.py
from pypika import Table, Query, Field, Column
from models import Timetable
from typing import List
from datetime import datetime as DateTime
import json
users, shared_timetable = Table('users'), Table('shared_timetable')

def get_timetable(user_id: int) -> str:
    """Get the timetable of the user with the given user_id"""
    query = Query.from_(users).select('timetable').where(users.id == user_id)
    return str(query)

def post_timetable(user_id: int, timetable: Timetable) -> str:
    """Update the timetable of the user with the given user_id"""
    timetable = json.dumps(timetable.model_dump())
    print(timetable)
    
    query = Query.update(users).set('timetable', timetable).where(users.id == user_id)
    return str(query)

def get_shared_timetable(code: str) -> str:
    """Get the timetable with the given code"""
    query = Query.from_(shared_timetable).select('*').where(shared_timetable.code == code)
    return str(query)

def post_shared_timetable(code: str, user_id: int, timetable, expiry: DateTime) -> str:
    """Post the timetable to the shared_timetable table"""
    query = Query.into(shared_timetable).columns('code', 'user_id', 'timetable', 'expiry').insert(code, user_id, json.dumps(timetable), expiry)
    return str(query)

def delete_shared_timetable(code: str) -> str:
    """Delete the timetable with the given code"""
    query = Query.from_(shared_timetable).delete().where(shared_timetable.code == code)
    return str(query)

File: backend-go/internal/db/timetable.go
package db

import (
	"context"
	"encoding/json"
	"errors"
	"time"

	_ "github.com/lib/pq"

	"github.com/LambdaIITH/Dashboard/backend/config"
	"github.com/LambdaIITH/Dashboard/backend/internal/schema"
)

// GetTimetable retrieves the timetable of the user with the given user_id.
func GetTimetable(ctx context.Context, userID int) (string, error) {
	query := "SELECT timetable FROM users WHERE id = $1"

	var result string
	err := config.DB.QueryRow(ctx, query, userID).Scan(&result)
	if err != nil {
		return "", err
	}
	return result, nil
}

// PostTimetable updates the timetable of the user with the given user_id.
func PostTimetable(ctx context.Context, userID int, timetable schema.Timetable) (map[string]interface{}, error) {
	timetableJSON, err := json.Marshal(timetable)
	if err != nil {
		return nil, err
	}

	query := "UPDATE users SET timetable = $1 WHERE id = $2"

	var result map[string]interface{}
	_, err = config.DB.Exec(ctx, query, string(timetableJSON), userID)
	if err != nil {
		return nil, err
	}
	return result, nil
}

// GetSharedTimetable retrieves the timetable with the given code.
func GetSharedTimetable(ctx context.Context, code string) (string, error) {
	query := "SELECT timetable, expiry FROM shared_timetable WHERE code = $1"
 
	var result string
	var expiry time.Time
	err := config.DB.QueryRow(ctx, query, code).Scan(&result, &expiry)

	if err != nil {
		if result == "" {
			return "", errors.New("no timetable found for the given code")
		}
		return "", err
	}

	if expiry.Before(time.Now()) {
		// Delete expired timetable
		deleteQuery := "DELETE FROM shared_timetable WHERE code = $1"
		_, err := config.DB.Exec(ctx, deleteQuery, code)
		if err != nil {
			return "", err
		}
		return "", errors.New("timetable has expired")
	}
	return result, nil
}

// PostSharedTimetable inserts a timetable into the shared_timetable table.
func PostSharedTimetable(ctx context.Context, code string, userID int, timetable interface{}, expiry time.Time) (map[string]interface{}, error) {
	timetableJSON, err := json.Marshal(timetable)
	if err != nil {
		return nil, err
	}

	query := "INSERT INTO shared_timetable (code, user_id, timetable, expiry) VALUES ($1, $2, $3, $4)"
	var result map[string]interface{}
	_, err = config.DB.Exec(ctx, query, code, userID, string(timetableJSON), expiry.Format(time.RFC3339))
	if err != nil {
		return nil, err
	}
	return result, nil
}

// DeleteSharedTimetable deletes the timetable with the given code.
func DeleteSharedTimetable(ctx context.Context, code string) (map[string]interface{}, error) {
	query := "DELETE FROM shared_timetable WHERE code = $1"
	var result map[string]interface{}
	_, err := config.DB.Exec(ctx, query, code)
	if err != nil {
		return nil, err
	}
	return result, nil
}
